File: Java/0148-sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) 
            return head;

        ListNode middle = getMiddle(head);
        ListNode left = head;
        ListNode right = middle.next;
        middle.next = null;

        return merge(sortList(left), sortList(right));
    }
    private ListNode getMiddle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head.next;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }
    private ListNode merge(ListNode left, ListNode right)
    {
        var head = new ListNode();
        var tail = head;
        while (left != null && right != null) {
            if (left.val < right.val) {
                tail.next = left;
                left = left.next;
            }
            else {
                tail.next = right;
                right = right.next;
            }
            tail = tail.next;
        }

        tail.next = left != null ? left : right;

        return head.next;
    }
}

File: Python/0148-sort-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None :
            return head 

        middle = self.get_middle(head)
        left = head
        right = middle.next
        middle.next = None
        return self.merge(self.sortList(left), self.sortList(right))

    def get_middle(self, head) :
        slow = head
        fast = head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        return slow

    def merge(self, left, right) :
        head = ListNode()
        curr = head
        while left and right :
            if left.val <= right.val :
                curr.next = left
                left = left.next
            else : 
                curr.next = right
                right = right.next
            curr = curr.next
        if left :
            curr.next = left
        else : 
            curr.next = right
        return head.next